### DISCUSSION TITLE
Tokenizer Incorrectly Splitting "M1M"

### DISCUSSION THREAD

**Discussion by jasondalycanpk** [explosion/spaCy#13360] [Help: Other Questions]
The tokenizer is incorrectly splitting the token M1M into M1 and M tokens.  See the following:

## How to reproduce the behaviour
Run the following code:
<pre>
import spacy

nlp = spacy.load("en_core_web_trf")
doc = nlp("hello there M1M 1M1")

for sent in doc.sents:
    for token in sent:
        print(token)
</pre>

This gives the following output:

<pre>
hello
there
M1
M
1M1
</pre>

## Your Environment
* Operating System: Debian Bookworm (12)
* Python Version Used: 3.11.x
* spaCy Version Used: spacy-3.5.4

**Comment by svlandeg** [comment]
Hi!

The tokenizer applies some heuristics, and in this case it's seeing "M" as a [unit](https://github.com/explosion/spaCy/blob/master/spacy/lang/char_classes.py#L265). You'd have the same behaviour when appending `K` or `mg`. The suffix rule responsible for this behaviour is [here](https://github.com/explosion/spaCy/blob/master/spacy/lang/punctuation.py#L28):

``` 
r"(?<=[0-9])(?:{u})".format(u=UNITS)
```

Like I said - these are heuristics that often help correctly tokenize texts where spaces are missing, but they might also result in false positive hits from time to time, as might be the case in your data. You could consider [customizing the tokenizer](https://spacy.io/usage/linguistic-features#native-tokenizers) for your use-case.
